### PR TITLE
Add Dataactions support for Roledefinition calls

### PR DESCRIPTION
**Pull request by darshanhs90**
Currently waiting for the following pr to get completed
https://github.com/Azure/azure-sdk-for-net/pull/4079 

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#making-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [ ] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

File: src/ResourceManager/Resources/Commands.Resources.Test/ScenarioTests/RoleDefinitionTests.cs [hunk 1]
```diff
@@ -70,15 +77,36 @@ public void RDRemoveScenario()
...
         }
 
         [Fact]
         [Trait(Category.AcceptanceType, Category.CheckIn)]
+        public void RDDataActionsNegativeTestCases()
+        {
+            ResourcesController.NewInstance.RunPsTest("Test-RDDataActionsNegativeTestCases");
+        }
+
+		[Fact]
```
Comment (maddieclayton) [inline src/ResourceManager/Resources/Commands.Resources.Test/ScenarioTests/RoleDefinitionTests.cs]: Fix formatting.

File: src/ResourceManager/Resources/Commands.Resources.Test/ScenarioTests/RoleDefinitionTests.ps1 [hunk 2]
```diff
@@ -191,6 +193,7 @@ function Test-RDRemove
     $badIdException = "RoleDefinitionDoesNotExist: The specified role definition with ID '" + $Rd.Id + "' does not exist."
     Assert-Throws { Remove-AzureRmRoleDefinition -Id $Rd.Id -Scope $scope -Force -PassThru} $badIdException
 
+    return
```
Comment (maddieclayton) [inline src/ResourceManager/Resources/Commands.Resources.Test/ScenarioTests/RoleDefinitionTests.ps1]: Why did you add this line?

File: src/ResourceManager/Resources/Commands.Resources/help/New-AzureRmRoleDefinition.md [hunk 3]
```diff
@@ -56,6 +56,9 @@ The input role definition MAY contain the following properties:
 
 1) NotActions: the set of operations that must be excluded from the Actions to determine the effective actions for the custom role.
 If there is a specific operation that you do not wish to grant access to in a custom role, it is convenient to use NotActions to exclude it, rather than specifying all operations other than that specific operation in Actions.
+2) DataActions: the set of data operations to which the custom role grants access.
```
Comment (maddieclayton) [inline src/ResourceManager/Resources/Commands.Resources/help/New-AzureRmRoleDefinition.md]: Please add this to the example below.

File: src/ResourceManager/Resources/Commands.Resources/help/Set-AzureRmRoleDefinition.md [hunk 4]
```diff
@@ -32,7 +32,7 @@ Set-AzureRmRoleDefinition -Role <PSRoleDefinition> [-DefaultProfile <IAzureConte
 The Set-AzureRmRoleDefinition cmdlet updates an existing custom role in Azure Role-Based Access Control.
 Provide the updated role definition as an input to the command as a JSON file or a PSRoleDefinition object.
 The role definition for the updated custom role MUST contain the Id and all other required properties of the role even if they are not updated: DisplayName, Description, Actions, AssignableScopes.
-NotActions is optional.
+NotActions, DataActions, NotDataActions are optional.
```
Comment (maddieclayton) [inline src/ResourceManager/Resources/Commands.Resources/help/Set-AzureRmRoleDefinition.md]: Add to example.

File: src/ResourceManager/Resources/Commands.Resources/Models.Authorization/AuthorizationClient.cs [hunk 5]
```diff
@@ -525,7 +518,7 @@ private static void ValidateRoleDefinition(PSRoleDefinition roleDefinition)
                 throw new ArgumentException(ProjectResources.InvalidAssignableScopes);
             }
 
-            if (roleDefinition.Actions == null || !roleDefinition.Actions.Any())
+            if ((roleDefinition.Actions == null || !roleDefinition.Actions.Any()) && (roleDefinition.DataActions == null || !roleDefinition.DataActions.Any()))
```
Comment (maddieclayton) [inline src/ResourceManager/Resources/Commands.Resources/Models.Authorization/AuthorizationClient.cs]: Will this cause any scripts to fail?

Comment (darshanhs90) [reply src/ResourceManager/Resources/Commands.Resources/Models.Authorization/AuthorizationClient.cs]: No.
earlier the condition was if they had no value actions,it used to error out.
but now the condition is that if they dont have any value for actions and also no value for datactions,it will error out

File: src/ResourceManager/Resources/Commands.Resources.Test/ScenarioTests/RoleDefinitionTests.ps1 (hunk 2 above)
Comment (darshanhs90) [reply src/ResourceManager/Resources/Commands.Resources.Test/ScenarioTests/RoleDefinitionTests.ps1]: reverting it

File: src/ResourceManager/Resources/Commands.Resources/help/New-AzureRmRoleDefinition.md (hunk 3 above)
Comment (darshanhs90) [reply src/ResourceManager/Resources/Commands.Resources/help/New-AzureRmRoleDefinition.md]: done

File: src/ResourceManager/Resources/Commands.Resources/help/Set-AzureRmRoleDefinition.md (hunk 4 above)
Comment (darshanhs90) [reply src/ResourceManager/Resources/Commands.Resources/help/Set-AzureRmRoleDefinition.md]: done

File: src/ResourceManager/Resources/Commands.Resources.Test/ScenarioTests/RoleDefinitionTests.cs [hunk 6]
```diff
@@ -35,42 +35,70 @@ public void RoleDefinitionCreateTests()
...
-		}
+        [Fact]
+        [Trait(Category.AcceptanceType, Category.CheckIn)]
+        public void RDCreateFromFile()
+        {
+            ResourcesController.NewInstance.RunPsTest("Test-RDCreateFromFile");
+        }
 
-		[Fact(Skip = "Unskip after service side change")]
+        [Fact(Skip = "Unskip after service side change")]
```
Comment (maddieclayton) [inline src/ResourceManager/Resources/Commands.Resources.Test/ScenarioTests/RoleDefinitionTests.cs]: Can you unskip this test now, or is this still waiting on other changes?

File: src/ResourceManager/Resources/Commands.Resources/help/New-AzureRmRoleDefinition.md [hunk 7]
```diff
@@ -56,29 +56,37 @@ The input role definition MAY contain the following properties:
...
+        \],
+        "DataActions":
+        \[
+            "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/read"
+        \],
+        "NotDataActions":
+        \[
+            "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write"
+        \],
+        "AssignableScopes": \["/subscriptions/4004a9fd-d58e-48dc-aeb2-4a4aec58606f"\]
```
Comment (maddieclayton) [inline src/ResourceManager/Resources/Commands.Resources/help/New-AzureRmRoleDefinition.md]: Change this to xxxx format - this looks like a real subscription.

File: src/ResourceManager/Resources/Commands.Resources/help/Set-AzureRmRoleDefinition.md [hunk 8]
```diff
@@ -45,9 +45,21 @@ Following is a sample updated role definition json for Set-AzureRmRoleDefinition
...
+        \],
+        "DataActions":
+        \[
+            "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/read"
+        \],
+        "NotDataActions":
+        \[
+            "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write"
+        \],
         "AssignableScopes": \["/subscriptions/4004a9fd-d58e-48dc-aeb2-4a4aec58606f"\]
```
Comment (maddieclayton) [inline src/ResourceManager/Resources/Commands.Resources/help/Set-AzureRmRoleDefinition.md]: Change this to xxxx format as well.

File: src/ResourceManager/Resources/Commands.Resources.Test/ScenarioTests/RoleDefinitionTests.cs (hunk 6 above)
Comment (darshanhs90) [reply src/ResourceManager/Resources/Commands.Resources.Test/ScenarioTests/RoleDefinitionTests.cs]: it is still waiting for the changes

**Comment by maddieclayton** [comment]
on demand here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/573/ (and 574)

File: src/ResourceManager/Resources/Commands.Resources/help/New-AzureRmRoleDefinition.md (hunk 7 above)
Comment (darshanhs90) [reply src/ResourceManager/Resources/Commands.Resources/help/New-AzureRmRoleDefinition.md]: updating it

File: src/ResourceManager/Resources/Commands.Resources/help/Set-AzureRmRoleDefinition.md (hunk 8 above)
Comment (darshanhs90) [reply src/ResourceManager/Resources/Commands.Resources/help/Set-AzureRmRoleDefinition.md]: updating it

**Comment by maddieclayton** [comment]
@darshanhs90 Once you fix the merge conflict that has come up, and fixed the text issue, I will merge this PR.

**Comment by maddieclayton** [comment]
on demand here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/583/ (and 585)
